File: library/src/test/java/de/westnordost/osmfeatures/FeatureTermIndexTest.java
package de.westnordost.osmfeatures;

import org.junit.Test;

import static de.westnordost.osmfeatures.MapEntry.mapOf;
import static de.westnordost.osmfeatures.TestUtils.assertEqualsIgnoreOrder;
import static de.westnordost.osmfeatures.TestUtils.listOf;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class FeatureTermIndexTest
{
    @Test public void copes_with_empty_collection()
    {
        FeatureTermIndex index = index();
        assertTrue(index.getAll("a").isEmpty());
    }


    @Test public void get_one_features_with_same_term()
    {
        Feature f1 = feature("a", "b");
        Feature f2 = feature("c");
        FeatureTermIndex index = index(f1, f2);
        assertEqualsIgnoreOrder(
                listOf(f1),
                index.getAll("b")
        );
    }

    @Test public void get_two_features_with_same_term()
    {
        Feature f1 = feature("a", "b");
        Feature f2 = feature("a", "c");
        FeatureTermIndex index = index(f1, f2);
        assertEqualsIgnoreOrder(
                listOf(f1, f2),
                index.getAll("a")
        );
    }

    @Test public void get_two_features_with_different_terms()
    {
        Feature f1 = feature("anything");
        Feature f2 = feature("anybody");
        FeatureTermIndex index = index(f1, f2);
        assertEqualsIgnoreOrder(
                listOf(f1, f2),
                index.getAll("any")
        );
        assertEqualsIgnoreOrder(
                listOf(f1),
                index.getAll("anyt")
        );
    }

    @Test public void dont_get_one_feature_twice()
    {
        Feature f1 = feature("something", "someone");
        FeatureTermIndex index = index(f1);
        assertEquals(listOf(f1), index.getAll("some"));
    }

    private static FeatureTermIndex index(Feature... features)
    {
        return new FeatureTermIndex(listOf(features), Feature::getTerms);
    }

    private static Feature feature(String... terms)
    {
        return new BaseFeature(
                "id",
                mapOf(),
                listOf(GeometryType.POINT),
                null,
                null,
                listOf("name"),
                listOf(terms),
                listOf(),
                listOf(),
                true,
                1.0,
                mapOf(),
                mapOf()
        );
    }
}

File: library/src/commonTest/kotlin/osmfeatures/FeatureTermIndexTest.kt
package de.westnordost.osmfeatures

import org.junit.Test
import osmfeatures.TestUtils.assertEqualsIgnoreOrder
import osmfeatures.TestUtils.listOf
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import osmfeatures.FeatureTermIndex

class FeatureTermIndexTest {
    @Test
    fun copes_with_empty_collection() {
        val index: FeatureTermIndex = index()
        assertTrue(index.getAll("a").isEmpty())
    }

    @Test
    fun get_one_features_with_same_term() {
        val f1: Feature = feature("a", "b")
        val f2: Feature = feature("c")
        val index: FeatureTermIndex = index(f1, f2)
        assertEqualsIgnoreOrder(
            listOf(f1),
            index.getAll("b")
        )
    }

    @Test
    fun get_two_features_with_same_term() {
        val f1: Feature = feature("a", "b")
        val f2: Feature = feature("a", "c")
        val index: FeatureTermIndex = index(f1, f2)
        assertEqualsIgnoreOrder(
            listOf(f1, f2),
            index.getAll("a")
        )
    }

    @Test
    fun get_two_features_with_different_terms() {
        val f1: Feature = feature("anything")
        val f2: Feature = feature("anybody")
        val index: FeatureTermIndex = index(f1, f2)
        assertEqualsIgnoreOrder(
            listOf(f1, f2),
            index.getAll("any")
        )
        assertEqualsIgnoreOrder(
            listOf(f1),
            index.getAll("anyt")
        )
    }

    @Test
    fun dont_get_one_feature_twice() {
        val f1: Feature = feature("something", "someone")
        val index: FeatureTermIndex = index(f1)
        assertEquals(listOf(f1), index.getAll("some"))
    }

    companion object {
        private fun index(vararg features: Feature): FeatureTermIndex {
            return FeatureTermIndex(features.toList()) { feature: Feature -> feature.terms }
        }

        private fun feature(vararg terms: String): Feature {
            return BaseFeature(
                "id",
                mapOf(),
                listOf(GeometryType.POINT),
                null,
                null,
                listOf("name"),
                terms.toList(),
                listOf(),
                listOf(),
                true,
                1.0f,
                false,
                mapOf(),
                mapOf()
            )
        }
    }
}
